File: takkernel/engine/src/main/java/com/atakmap/map/layer/raster/osm/OSMDroidInfo.java
package com.atakmap.map.layer.raster.osm;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import com.atakmap.coremap.io.DatabaseInformation;
import com.atakmap.coremap.io.IOProviderFactory;
import com.atakmap.coremap.log.Log;
import com.atakmap.database.CursorIface;
import com.atakmap.database.DatabaseIface;
import com.atakmap.database.Databases;
import com.atakmap.database.QueryIface;
import com.atakmap.map.projection.WebMercatorProjection;

import java.io.File;

public final class OSMDroidInfo {
    public enum BoundsDiscovery {
        /**
         * Full bounds discovery will be performed. This will require a full
         * table scan. This is the slowest option.
         */
        Full,
        /**
         * Bounds are computed based on the minimum and maximum keys for the
         * minimum level. This will be accurate for datasets with a single,
         * contiguous extent that is rectangular in shape. The grid min/max Y
         * will not be accurate in the event that the contained tiles are non-
         * contiguous or form a non-rectangular extent.
         */
        Quick,
        /**
         * Bounds computation is skipped altogether. Grid extents are populated
         * with the maximum possible values.
         */
        Skip,
    };

    public int srid;
    public int minLevel;
    public int maxLevel;
    public int minLevelGridMinX;
    public int minLevelGridMinY;
    public int minLevelGridMaxX;
    public int minLevelGridMaxY;
    public String provider;
    public int tileWidth;
    public int tileHeight;
    public int gridZeroWidth;
    public int gridZeroHeight;
 
    /**************************************************************************/

    public static OSMDroidInfo get(String path, BoundsDiscovery bounds, DatabaseIface[] returnRef) {
        DatabaseIface database;
        
        // try spatialite first
        database = null;
        try {
            database = IOProviderFactory.createDatabase(new File(path), DatabaseInformation.OPTION_READONLY);
            final OSMDroidInfo retval = get(database, bounds);
            if(retval != null) {
                if(returnRef != null) {
                    returnRef[0] = database;
                    database = null;
                }
                database = null;
                return retval;
            }
        } catch(Throwable ignored) {
        } finally {
            if(database != null)
                database.close();
        }
        
        return null;
    }

    public static OSMDroidInfo get(DatabaseIface database, BoundsDiscovery bounds) {
        try {
            if(!OSMUtils.isOSMDroidSQLite(database))
                return null;
            
            OSMDroidInfo info = new OSMDroidInfo();
            
            // OSM Droid SQLite
            CursorIface result;

            final long minKey;
            
            // min key discovery
            result = null;
            try {
                result = database.query("SELECT min(key) FROM tiles",
                        null);
                if (result.moveToNext() && !result.isNull(0)) {
                    minKey = result.getLong(0);
                    info.minLevel = OSMUtils.getOSMDroidSQLiteZoomLevel(minKey);
                } else {
                    // no tiles present in database
                    return null;
                }
            } finally {
                if (result != null)
                    result.close();
            }

            // max key discovery
            result = null;
            try {
                result = database.query("SELECT max(key) FROM tiles",
                        null);
                if (result.moveToNext() && !result.isNull(0)) {
                    info.maxLevel = OSMUtils.getOSMDroidSQLiteZoomLevel(result.getLong(0));
                } else {
                    throw new IllegalStateException();
                }
            } finally {
                if (result != null)
                    result.close();
            }

            // obtain the SRID
            info.srid = WebMercatorProjection.INSTANCE.getSpatialReferenceID();
            info.gridZeroWidth = 1;
            info.gridZeroHeight = 1;
            if (Databases.getColumnNames(database, "ATAK_metadata") != null) {
                result = null;
                try {
                    result = database.query(
                            "SELECT value FROM ATAK_metadata WHERE key = \'srid\'", null);
                    if (result.moveToNext()) {
                        info.srid = Integer.parseInt(result.getString(0));
                        if(info.srid == 4326) {
                            info.gridZeroWidth = 2;
                            info.gridZeroHeight = 1;
                        }
                    }
                } catch (Exception ignored) {
                    // quietly ignore
                } finally {
                    if (result != null)
                        result.close();
                }
            }
            
            // bounds discovery
            switch(bounds) {
                case Full :
                {
                    // find the min/max grid x -- this can be done with 2 single
                    // queries. assign initial min/max grid y.
                    result = null;
                    try {
                        result = database.compileQuery("SELECT min(key) FROM tiles WHERE key <= ?");
                        ((QueryIface)result).bind(1, OSMUtils.getOSMDroidSQLiteMaxIndex(info.minLevel));
                        
                        result.moveToNext();
                        info.minLevelGridMinX = OSMUtils.getOSMDroidSQLiteTileX(result.getLong(0));
                        
                        final int y = OSMUtils.getOSMDroidSQLiteTileY(result.getLong(0));
                        info.minLevelGridMinY = y;
                        info.minLevelGridMaxY = y;
                    } finally {
                        if (result != null)
                            result.close();
                    }
                    
                    result = null;
                    try {
                        result = database.compileQuery("SELECT max(key) FROM tiles WHERE key <= ?");
                        ((QueryIface)result).bind(1, OSMUtils.getOSMDroidSQLiteMaxIndex(info.minLevel));
                        
                        result.moveToNext();
                        info.minLevelGridMaxX = OSMUtils.getOSMDroidSQLiteTileX(result.getLong(0));
                        
                        final int y = OSMUtils.getOSMDroidSQLiteTileY(result.getLong(0));
                        if(y < info.minLevelGridMinY)
                            info.minLevelGridMinY = y;
                        if(y > info.minLevelGridMaxY)
                            info.minLevelGridMaxY = y;
                    } finally {
                        if (result != null)
                            result.close();
                    }
                    
                    // this method is quicker than brute force, but is not
                    // logarithmic time. iterate between the min/max x values
                    // doing min/max y discovery on each grid column, shrinking
                    // the search based on the current min/max y values
                    for(int x = info.minLevelGridMinX; x <= info.minLevelGridMaxX; x++) {
                        result = null;
                        try {
                            result = database.compileQuery("SELECT min(key) FROM tiles WHERE key >= ? AND key < ?");
                            ((QueryIface)result).bind(1, OSMUtils.getOSMDroidSQLiteIndex(info.minLevel, x, 0));
                            ((QueryIface)result).bind(2, OSMUtils.getOSMDroidSQLiteIndex(info.minLevel, x, info.minLevelGridMinY));
                            
                            if(result.moveToNext() && !result.isNull(0)) {
                                final int y = OSMUtils.getOSMDroidSQLiteTileY(result.getLong(0));
                                if(y < info.minLevelGridMinY)
                                    info.minLevelGridMinY = y;
                            }
                        } finally {
                            if (result != null)
                                result.close();
                        }
                        
                        result = null;
                        try {
                            result = database.compileQuery("SELECT max(key) FROM tiles WHERE key > ? AND key <= ?");
                            ((QueryIface)result).bind(1, OSMUtils.getOSMDroidSQLiteIndex(info.minLevel, x, info.minLevelGridMaxY));
                            ((QueryIface)result).bind(2, OSMUtils.getOSMDroidSQLiteIndex(info.minLevel, x, (1<<info.minLevel)-1));
                            
                            if(result.moveToNext() && !result.isNull(0)) {
                                final int y = OSMUtils.getOSMDroidSQLiteTileY(result.getLong(0));
                                if(y > info.minLevelGridMaxY)
                                    info.minLevelGridMaxY = y;
                            }
                        } finally {
                            if (result != null)
                                result.close();
                        }
                    }
                    break;
                }
                case Quick :
                {
                    result = null;
                    try {
                        final int mask = ~(0xFFFFFFFF<<info.minLevel);
                        result = database.compileQuery("SELECT min(key) FROM tiles WHERE key <= ?");
                        ((QueryIface)result).bind(1, OSMUtils.getOSMDroidSQLiteMaxIndex(info.minLevel));
                        
                        result.moveToNext();
                        info.minLevelGridMinX = OSMUtils.getOSMDroidSQLiteTileX(result.getLong(0));
                        info.minLevelGridMinY = OSMUtils.getOSMDroidSQLiteTileY(result.getLong(0));
                    } finally {
                        if (result != null)
                            result.close();
                    }
                    
                    result = null;
                    try {
                        final int mask = ~(0xFFFFFFFF<<info.minLevel);
                        result = database.compileQuery("SELECT max(key) FROM tiles WHERE key <= ?");
                        ((QueryIface)result).bind(1, OSMUtils.getOSMDroidSQLiteMaxIndex(info.minLevel));
                        
                        result.moveToNext();
                        info.minLevelGridMaxX = OSMUtils.getOSMDroidSQLiteTileX(result.getLong(0));
                        info.minLevelGridMaxY = OSMUtils.getOSMDroidSQLiteTileY(result.getLong(0));
                    } finally {
                        if (result != null)
                            result.close();
                    }
                    break;
                }
                case Skip :
                    info.minLevelGridMinX = 0;
                    info.minLevelGridMinY = 0;
                    info.minLevelGridMaxX = (info.gridZeroWidth << info.minLevel);
                    info.minLevelGridMaxY = (info.gridZeroHeight << info.minLevel);
                    break;
                default :
                    throw new IllegalArgumentException();
            }

            // obtain provider and tile dimensions
            info.provider = null;
            info.tileWidth = 256;
            info.tileHeight = 256;
            
            result = null;
            try {
                result = database.query("SELECT provider, tile FROM tiles LIMIT 1",
                        null);
                if (result.moveToNext()) {
                    info.provider = result.getString(0);
                    
                    byte[] blob = result.getBlob(1);
                    BitmapFactory.Options opts = new BitmapFactory.Options();
                    opts.inJustDecodeBounds = true;
                    Bitmap tile = null;
                    try {
                        tile = BitmapFactory.decodeByteArray(blob, 0, blob.length, opts);
                        info.tileWidth = opts.outWidth;
                        info.tileHeight = opts.outHeight;
                    } finally {
                        if(tile != null)
                            tile.recycle();
                    }
                } else {
                    throw new IllegalStateException();
                }
            } finally {
                if (result != null)
                    result.close();
            }
            
            return info;
        } catch(Exception e) {
            Log.w("OSMDroidInfo", "Unexpected error parsing info", e);
            return null;
        }
    }    
}

File: takkernel/engine/src/main/cpp/formats/osmdroid/OSMDroidInfo.cpp
#ifdef MSVC
#include "formats/osmdroid/OSMDroidInfo.h"

#include <vector>

#include "db/DatabaseFactory.h"
#include "db/Query.h"
#include "port/STLVectorAdapter.h"
#include "raster/osm/OSMUtils.h"
#include "renderer/BitmapFactory2.h"

using namespace TAK::Engine::Formats::OSMDroid;

using namespace TAK::Engine::DB;
using namespace TAK::Engine::Util;

TAKErr TAK::Engine::Formats::OSMDroid::OSMDroidInfo_get(OSMDroidInfo *value, DatabasePtr *dbref, const char* path, const BoundsDiscovery bounds) NOTHROWS
{
    TAKErr code(TE_Ok);

    DatabasePtr database(nullptr, nullptr);
    code = DatabaseFactory_create(database, DatabaseInformation(path, nullptr, DATABASE_OPTIONS_READONLY));
    TE_CHECKRETURN_CODE(code);

    code = OSMDroidInfo_get(value, *database, bounds);
    TE_CHECKRETURN_CODE(code);

    if(dbref)
        (*dbref) = std::move(database);
    
    return code;
}

TAKErr TAK::Engine::Formats::OSMDroid::OSMDroidInfo_get(OSMDroidInfo* value, Database2& database, const BoundsDiscovery bounds) NOTHROWS
{
    TAKErr code(TE_Ok);
    if (!atakmap::raster::osm::OSMUtils::isOSMDroidSQLite(database))
        return TE_InvalidArg;
            
    // OSM Droid SQLite
    QueryPtr result(nullptr, nullptr);

    int64_t minKey;
            
    // min key discovery
    result.reset();
    code = database.query(result, "SELECT min(key) FROM tiles");
    do {
        code = result->moveToNext();
        TE_CHECKBREAK_CODE(code);
        bool isNull;
        code = result->isNull(&isNull, 0);
        TE_CHECKBREAK_CODE(code);
        if (isNull)
            return TE_InvalidArg;
        
        code = result->getLong(&minKey, 0);
        TE_CHECKBREAK_CODE(code);

        value->minLevel = atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteZoomLevel(minKey);
    } while (false);
    TE_CHECKRETURN_CODE(code);

    // max key discovery
    result.reset();
    code = database.query(result, "SELECT max(key) FROM tiles");
    do {
        code = result->moveToNext();
        TE_CHECKBREAK_CODE(code);
        bool isNull;
        code = result->isNull(&isNull, 0);
        TE_CHECKBREAK_CODE(code);
        if (isNull)
            return TE_InvalidArg;
        
        int64_t maxKey;
        code = result->getLong(&maxKey, 0);
        TE_CHECKBREAK_CODE(code);

        value->maxLevel = atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteZoomLevel(maxKey);
    } while (false);
    TE_CHECKRETURN_CODE(code);

    // obtain the SRID
    value->srid = 3857;
    value->gridZeroWidth = 1;
    value->gridZeroHeight = 1;

    std::vector<TAK::Engine::Port::String> atakMetadataColumnNames;
    TAK::Engine::Port::STLVectorAdapter<TAK::Engine::Port::String> atakMetadataColumnNames_a(atakMetadataColumnNames);
    if (Databases_getColumnNames(atakMetadataColumnNames_a, database, "ATAK_metadata") == TE_Ok && !atakMetadataColumnNames.empty()) {
        result.reset();
        do {
            if (database.query(result, "SELECT value FROM ATAK_metadata WHERE key = \'srid\'") != TE_Ok)
                break;
            if (result->moveToNext() != TE_Ok)
                break;
            const char *srid;
            if (result->getString(&srid, 0) != TE_Ok)
                break;
            if (TAK::Engine::Port::String_parseInteger(&value->srid, srid) != TE_Ok)
                break;
            if (value->srid == 4326) {
                value->gridZeroWidth = 2;
                value->gridZeroHeight = 1;
            }
        } while (false);
    }
            
    // bounds discovery
    switch(bounds) {
        case TEBD_Full :
        {
            // find the min/max grid x -- this can be done with 2 single
            // queries. assign initial min/max grid y.
            result.reset();
            {
                code = database.compileQuery(result, "SELECT min(key) FROM tiles WHERE key <= ?");
                TE_CHECKRETURN_CODE(code);
                code = result->bindLong(1, atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteMaxIndex(value->minLevel));
                TE_CHECKRETURN_CODE(code);

                code = result->moveToNext();
                TE_CHECKRETURN_CODE(code);
                int64_t tileKey;
                code = result->getLong(&tileKey, 0);
                value->minLevelGridMinX = atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteTileX(tileKey);

                const int y = atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteTileY(tileKey);
                value->minLevelGridMinY = y;
                value->minLevelGridMaxY = y;
            }
                    
            result.reset();
            {
                code = database.compileQuery(result, "SELECT max(key) FROM tiles WHERE key <= ?");
                TE_CHECKRETURN_CODE(code);
                code = result->bindLong(1, atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteMaxIndex(value->minLevel));
                TE_CHECKRETURN_CODE(code);

                code = result->moveToNext();
                TE_CHECKRETURN_CODE(code);
                int64_t tileKey;
                code = result->getLong(&tileKey, 0);
                TE_CHECKRETURN_CODE(code);
                value->minLevelGridMaxX = atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteTileX(tileKey);

                const int y = atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteTileY(tileKey);
                if (y < value->minLevelGridMinY)
                    value->minLevelGridMinY = y;
                if (y > value->minLevelGridMaxY)
                    value->minLevelGridMaxY = y;
            }
                    
            // this method is quicker than brute force, but is not
            // logarithmic time. iterate between the min/max x values
            // doing min/max y discovery on each grid column, shrinking
            // the search based on the current min/max y values
            for(int x = value->minLevelGridMinX; x <= value->minLevelGridMaxX; x++) {
                result.reset();
                do {
                    code = database.compileQuery(result, "SELECT min(key) FROM tiles WHERE key >= ? AND key < ?");
                    TE_CHECKRETURN_CODE(code);
                    result->bindLong(1, atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteIndex(value->minLevel, x, 0));
                    result->bindLong(2, atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteIndex(value->minLevel, x, value->minLevelGridMinY));
                            
                    if (result->moveToNext() != TE_Ok)
                        break;
                    bool isNull = true;
                    if (result->isNull(&isNull, 0) != TE_Ok || isNull)
                        break;

                    int64_t tileKey;
                    if (result->getLong(&tileKey, 0) != TE_Ok)
                        break;
                    const int y = atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteTileY(tileKey);
                    if(y < value->minLevelGridMinY)
                        value->minLevelGridMinY = y;
                } while (false);
                        
                result.reset();
                do {
                    code = database.compileQuery(result, "SELECT max(key) FROM tiles WHERE key > ? AND key <= ?");
                    TE_CHECKBREAK_CODE(code);
                    result->bindLong(1, atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteIndex(value->minLevel, x, value->minLevelGridMaxY));
                    result->bindLong(2, atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteIndex(value->minLevel, x, (1<<value->minLevel)-1));

                    if (result->moveToNext() != TE_Ok)
                        break;
                    bool isNull = true;
                    if (result->isNull(&isNull, 0) != TE_Ok || isNull)
                        break;

                    int64_t tileKey;
                    if (result->getLong(&tileKey, 0) != TE_Ok)
                        break;
                    const int y = atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteTileY(tileKey);
                    if(y > value->minLevelGridMaxY)
                        value->minLevelGridMaxY = y;
                } while (false);
            }
            break;
        }
        case TEBD_Quick :
        {
            result.reset();
            {
                const int mask = ~(0xFFFFFFFF<<value->minLevel);
                code = database.compileQuery(result, "SELECT min(key) FROM tiles WHERE key <= ?");
                TE_CHECKRETURN_CODE(code);
                code = result->bindLong(1, atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteMaxIndex(value->minLevel));
                TE_CHECKRETURN_CODE(code);
                        
                code = result->moveToNext();
                TE_CHECKRETURN_CODE(code);
                
                int64_t tileKey;
                code = result->getLong(&tileKey, 0);
                TE_CHECKRETURN_CODE(code);

                value->minLevelGridMinX = atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteTileX(tileKey);
                value->minLevelGridMinY = atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteTileY(tileKey);
            }
                    
            result.reset();
            {
                const int mask = ~(0xFFFFFFFF<<value->minLevel);
                code = database.compileQuery(result, "SELECT max(key) FROM tiles WHERE key <= ?");
                TE_CHECKRETURN_CODE(code);
                code = result->bindLong(1, atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteMaxIndex(value->minLevel));
                TE_CHECKRETURN_CODE(code);
                        
                code = result->moveToNext();
                TE_CHECKRETURN_CODE(code);

                int64_t tileKey;
                code = result->getLong(&tileKey, 0);
                TE_CHECKRETURN_CODE(code);

                value->minLevelGridMaxX = atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteTileX(tileKey);
                value->minLevelGridMaxY = atakmap::raster::osm::OSMUtils::getOSMDroidSQLiteTileY(tileKey);
            }
            break;
        }
        case TEBD_Skip :
            value->minLevelGridMinX = 0;
            value->minLevelGridMinY = 0;
            value->minLevelGridMaxX = (value->gridZeroWidth << value->minLevel);
            value->minLevelGridMaxY = (value->gridZeroHeight << value->minLevel);
            break;
        default :
            return TE_InvalidArg;
    }

    // obtain provider and tile dimensions
    value->provider = nullptr;
    value->tileWidth = 256;
    value->tileHeight = 256;
            
    result.reset();
    {
        code = database.query(result, "SELECT provider, tile FROM tiles LIMIT 1");
        TE_CHECKRETURN_CODE(code);
        if (result->moveToNext() == TE_Ok) {
            const char* provider;
            code = result->getString(&provider, 0);
            TE_CHECKRETURN_CODE(code);
            value->provider = provider;
                    

            const uint8_t *blob;
            std::size_t blobLen;
            code = result->getBlob(&blob, &blobLen, 1);
            TE_CHECKRETURN_CODE(code);

            TAK::Engine::Renderer::BitmapPtr tile(nullptr, nullptr);
            code = TAK::Engine::Renderer::BitmapFactory2_decode(tile, blob, blobLen, nullptr);
            TE_CHECKRETURN_CODE(code);

            value->tileWidth = tile->getWidth();
            value->tileHeight = tile->getHeight();
        } else {
            return TE_IllegalState;
        }
    }
            
    return code;
}
#endif